apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1358 @@
+        _stakeBalance(_from, _from, _amount);
+    }
+
+    /**
+    * @notice Unstake `@tokenAmount(self.collateralToken(): address, _amount)` tokens from `msg.sender`
+    * @param _amount Number of collateral tokens to be unstaked
+    */
+    function unstake(uint256 _amount) external {
+        require(_amount > 0, ERROR_INVALID_UNSTAKE_AMOUNT);
+        _unstakeBalance(msg.sender, _amount);
+    }
+
+    /**
+    * @notice Schedule a new action
+    * @param _context Link to a human-readable text giving context for the given action
+    * @param _script Action script to be executed
+    */
+    function schedule(bytes _context, bytes _script) external onlySigner(msg.sender) {
+        _createAction(msg.sender, _context, _script);
+    }
+
+    /**
+    * @notice Execute action #`_actionId`
+    * @dev It only executes non-challenged actions after the challenge period or actions that were disputed but ruled in favor of the submitter
+    * @param _actionId Identification number of the action to be executed
+    */
+    function execute(uint256 _actionId) external {
+        (Action storage action, Setting storage setting) = _getActionAndSetting(_actionId);
+        require(_canExecute(action), ERROR_CANNOT_EXECUTE_ACTION);
+
+        if (action.state == ActionState.Scheduled) {
+            _unlockBalance(action.submitter, setting.collateralAmount);
+        }
+        action.state = ActionState.Executed;
+        runScript(action.script, new bytes(0), _getScriptExecutionBlacklist());
+        emit ActionExecuted(_actionId);
+    }
+
+    /**
+    * @notice Cancel action #`_actionId`
+    * @dev It only cancels non-challenged actions or actions that were disputed but ruled in favor of the submitter
+    * @param _actionId Identification number of the action to be cancelled
+    */
+    function cancel(uint256 _actionId) external {
+        (Action storage action, Setting storage setting) = _getActionAndSetting(_actionId);
+        require(_canCancel(action), ERROR_CANNOT_CANCEL_ACTION);
+
+        address submitter = action.submitter;
+        require(msg.sender == submitter, ERROR_SENDER_NOT_ALLOWED);
+
+        if (action.state == ActionState.Scheduled) {
+            _unlockBalance(submitter, setting.collateralAmount);
+        }
+        action.state = ActionState.Cancelled;
+        emit ActionCancelled(_actionId);
+    }
+
+    /**
+    * @notice Challenge an action #`_actionId` with a settlement offer of `@tokenAmount(self.collateralToken(): address, _settlementOffer)`
+    * @param _actionId Identification number of the action being challenged
+    * @param _settlementOffer Amount of collateral tokens the challenger would accept for resolving the dispute without involving the arbitrator
+    * @param _context Link to a human-readable text giving context for the challenge
+    */
+    function challengeAction(uint256 _actionId, uint256 _settlementOffer, bytes _context) external authP(CHALLENGE_ROLE, arr(_actionId)) {
```

Comment:
Allowing only holders of a certain token to challenge actions could be a pretty common configuration for Agreements.

Since we require `msg.sender` to be the actual originator of the intent (i.e. cannot use a TokenManager forwarder as we need to `transferFrom` the account that challenges), in order for us to limit who can challenge based on a token balance we'd need to use a parametrized role with an ACL oracle or use the same approach as in `canSign`.

If we take the approach of having a custom permission here as well, we will need to decide whether it would be completely different from the signing permission, allowing the token and the amount that you need to hold to be different

Reply:
Interesting, let's try to have a separate balance condition here then :)